server/src/main/java/org/apache/kafka/server/DelegationTokenManager.java

```diff
@@ -0,0 +1,124 @@
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.function.Function;
+
+import javax.crypto.Mac;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.SecretKeySpec;
+
+public class DelegationTokenManager {
+    private static final String DEFAULT_HMAC_ALGORITHM = "HmacSHA512";
+    public static final long ERROR_TIMESTAMP = -1;
+
+    private final DelegationTokenCache tokenCache;
+    private final SecretKey secretKey;
+
+    public DelegationTokenManager(DelegationTokenManagerConfigs config, DelegationTokenCache tokenCache) {
+        this.tokenCache = tokenCache;
+
+        byte[] keyBytes = config.tokenAuthEnabled() ? config.delegationTokenSecretKey().value().getBytes(StandardCharsets.UTF_8) : null;
+        if (keyBytes == null || keyBytes.length == 0) {
+            this.secretKey = null;
+        } else {
+            this.secretKey = createSecretKey(keyBytes);
+        }
+    }
+
+    private static SecretKey createSecretKey(byte[] keyBytes) {
+        return new SecretKeySpec(keyBytes, DEFAULT_HMAC_ALGORITHM);
+    }
+
+    public static byte[] createHmac(String tokenId, SecretKey secretKey) {
+        try {
+            Mac mac = Mac.getInstance(DEFAULT_HMAC_ALGORITHM);
+            mac.init(secretKey);
+            return mac.doFinal(tokenId.getBytes(StandardCharsets.UTF_8));
+        } catch (InvalidKeyException e) {
+            throw new IllegalArgumentException("Invalid key to HMAC computation", e);
+        } catch (Exception e) {
+            throw new RuntimeException("Error while creating HMAC", e);
+        }
+    }
+
+    private Map<String, ScramCredential> prepareScramCredentials(String hmacString) throws NoSuchAlgorithmException {
+        Map<String, ScramCredential> scramCredentialMap = new HashMap<>();
+        for (ScramMechanism mechanism : ScramMechanism.values()) {
+            ScramFormatter formatter = new ScramFormatter(mechanism);
+            scramCredentialMap.put(mechanism.mechanismName(), formatter.generateCredential(hmacString, mechanism.minIterations()));
+        }
+        return scramCredentialMap;
+    }
+
+    public void updateToken(DelegationToken token) throws NoSuchAlgorithmException {
+        String hmacString = token.hmacAsBase64String();
+        Map<String, ScramCredential> scramCredentialMap = prepareScramCredentials(hmacString);
+        tokenCache.updateCache(token, scramCredentialMap);
+    }
+
+    public DelegationToken getDelegationToken(TokenInformation tokenInfo) {
+        byte[] hmac = createHmac(tokenInfo.tokenId(), secretKey);
+        return new DelegationToken(tokenInfo, hmac);
+    }
+
+    public void removeToken(String tokenId) {
+        tokenCache.removeCache(tokenId);
+    }
+
+    public List<DelegationToken> getTokens(java.util.function.Predicate<TokenInformation> filterToken) {
```

Comment:
We can import `java.util.function.Predicate`

Reply:
Updated it. Thank you.